src/tools/compiletest/src/util/edit_distance.rs

```diff
@@ -0,0 +1,239 @@
+//! Edit distance taken from rustc's `rustc_span::edit_distance`.
```

Comment:
I don't feel comfortable duplicating this. Do we have any alternative options?

Reply:
We can also just not provide any suggestions tbh, so this duplication won't be needed in the first place.

Reply:
That could work. Perhaps we could search for a lightweight (to avoid increasing the build time of tidy) and well-maintained dependency to use.

Reply:
> That could work. Perhaps we could search for a lightweight (to avoid increasing the build time of tidy) and well-maintained dependency to use.

In the grand scheme of things, we should probably have a edit distance and suggestion crate under rust-lang so rustc and cargo can share the same edit distance and suggestion logic. That way, tidy and compiletest and other tools could also use the crate if suggestions might be helpful to contributors.

Reply:
https://crates.io/crates/levenshtein_automata seems to be a good candidate since it is actively used by the [tantivy](https://github.com/quickwit-oss/tantivy) search engine.

Reply:
> https://crates.io/crates/levenshtein_automata seems to be a good candidate since it is actively used by the [tantivy](https://github.com/quickwit-oss/tantivy) search engine.

Yep. I'll hold off adding any suggestions in this particular PR for now, it can always be added easily later.

Reply:
> In the grand scheme of things, we should probably have a edit distance and suggestion crate under rust-lang so rustc and cargo can share the same edit distance and suggestion logic. That way, tidy and compiletest and other tools could also use the crate if suggestions might be helpful to contributors.

That's a good idea if we have multiple tools requiring suggestions. However, for now, I think it might be overkill (as having an additional tool increases the cost of maintenance).

> We can also just not provide any suggestions tbh

We can consider following this approach for now. If a directive is invalid, we can simply include the location of all available directives in the error message (e.g., go to /x/y/z/t.rs and check DIRECTIVES variable).